Source/WebKit/Shared/WebBackForwardListFrameItem.cpp

```diff
@@ -32,47 +32,53 @@
+Ref<WebBackForwardListFrameItem> WebBackForwardListFrameItem::create(WebBackForwardListItem& item, WebBackForwardListFrameItem* parentItem, Ref<FrameState>&& frameState)
 {
     return adoptRef(*new WebBackForwardListFrameItem(item, parentItem, WTFMove(frameState)));
 }
 
-WebBackForwardListFrameItem::WebBackForwardListFrameItem(WebBackForwardListItem* item, WebBackForwardListFrameItem* parentItem, Ref<FrameState>&& frameState)
+WebBackForwardListFrameItem::WebBackForwardListFrameItem(WebBackForwardListItem& item, WebBackForwardListFrameItem* parentItem, Ref<FrameState>&& frameState)
     : m_backForwardListItem(item)
     , m_frameState(WTFMove(frameState))
     , m_parent(parentItem)
 {
-    auto result = allItems().add(*m_frameState->identifier, *this);
+    m_frameState->itemID = item.identifier();
+    auto result = allItems().add({ *m_frameState->frameItemID, *m_frameState->itemID }, *this);
     ASSERT_UNUSED(result, result.isNewEntry);
     for (auto& child : std::exchange(m_frameState->children, { }))
         m_children.append(WebBackForwardListFrameItem::create(item, this, WTFMove(child)));
 }
 
 WebBackForwardListFrameItem::~WebBackForwardListFrameItem()
 {
-    ASSERT(allItems().get(*m_frameState->identifier) == this);
-    allItems().remove(*m_frameState->identifier);
+    ASSERT(allItems().get({ *m_frameState->frameItemID, *m_frameState->itemID }) == this);
+    allItems().remove({ *m_frameState->frameItemID, *m_frameState->itemID });
 }
 
-HashMap<BackForwardItemIdentifier, WeakRef<WebBackForwardListFrameItem>>& WebBackForwardListFrameItem::allItems()
+HashMap<std::pair<BackForwardFrameItemIdentifier, BackForwardItemIdentifier>, WeakRef<WebBackForwardListFrameItem>>& WebBackForwardListFrameItem::allItems()
 {
-    static MainThreadNeverDestroyed<HashMap<BackForwardItemIdentifier, WeakRef<WebBackForwardListFrameItem>>> items;
+    static MainThreadNeverDestroyed<HashMap<std::pair<BackForwardFrameItemIdentifier, BackForwardItemIdentifier>, WeakRef<WebBackForwardListFrameItem>>> items;
     return items;
 }
 
-WebBackForwardListFrameItem* WebBackForwardListFrameItem::itemForID(BackForwardItemIdentifier identifier)
+WebBackForwardListFrameItem* WebBackForwardListFrameItem::itemForID(BackForwardItemIdentifier itemID, BackForwardFrameItemIdentifier frameItemID)
 {
-    return allItems().get(identifier);
+    return allItems().get({ frameItemID, itemID });
 }
 
 std::optional<FrameIdentifier> WebBackForwardListFrameItem::frameID() const
 {
     return m_frameState->frameID;
 }
 
-BackForwardItemIdentifier WebBackForwardListFrameItem::identifier() const
+const String& WebBackForwardListFrameItem::url() const
 {
-    return *m_frameState->identifier;
+    return m_frameState->urlString;
+}
+
+BackForwardFrameItemIdentifier WebBackForwardListFrameItem::identifier() const
+{
+    return *m_frameState->frameItemID;
```

Comment:
Can we express in the type that this is always present without dereferencing an optional?